lib/utils/device/live.ts

```diff
@@ -162,11 +165,43 @@ export class LivepushManager {
 					eventQueue.push(changedPath);
 					this.getDebouncedEventHandler(serviceName)();
 				};
+
+				// Get list of paths for file watcher to ignore using dockerignores
+				const ignored = ['.git'];
+				const serviceDirs = this.deployOpts.multiDockerignore
+					? await getServiceDirsFromComposition(
+							this.deployOpts.source,
+							this.composition,
+					  )
+					: {};
+				const { dockerignoreFiles } = await filterFilesWithDockerignore(
+					this.deployOpts.source,
+					serviceDirs,
+				);
+				let dockerignoreContent = '';
+				let ignorePath = '';
+				for (const file of dockerignoreFiles) {
+					try {
+						dockerignoreContent = await fs.readFile(file.filePath, 'utf8');
+					} catch (err) {
+						this.logger.logError(
+							`Error reading .dockerignore file "${file.filePath}":\n${err}`,
+						);
+						throw err;
+					}
+					for (const line of dockerignoreContent.split(/\r?\n/)) {
+						ignorePath = line.trim();
+						if (!_.isEmpty(ignorePath)) {
+							ignored.push(path.join(path.dirname(file.relPath), ignorePath));
```

Comment:
Unfortunately it may not be this simple because `chokidar` uses `anymatch` as the file pattern matcher:
https://github.com/micromatch/anymatch (See also: https://www.npmjs.com/package/chokidar)

... whereas the `.dockerignore` files use Docker's own pattern matcher, which you can read more about here:
https://github.com/balena-io-modules/dockerignore/#dockerignore

For example, the dockerignore format allows adding pattern match exceptions with the exclamation mark, whereas anymatch has no such concept. The anymatch documentation says it has "bash parity", so presumably (?) it understands bash "wildcards" like `?` and ranges like `a{b,c}`, which are not understood by dockerignore. Also, the anymatch documentation says it does not support the backslash as path separator on Windows, whereas dockerignore does.

I suspect it would be properly painful to try and map the differences between pattern matchers.

Therefore, instead of this approach, what comes to mind is two alternatives:
* Provide an exhaustive, pre-filtered list of files to watch to chokidar. I haven't used chokidar much, but [its documentation] reads:
```
// Initialize watcher.
const watcher = chokidar.watch('file, dir, glob, or array', ...
```
So the idea here would be, instead of `chokidar.watch('.', ...` as the CLI currently does, to use `chokidar.watch(fileList, ...` where `fileList` is a list of every single file to watch, pre-filtered with dockerignore.

* Another approach would to be provide a function as the value of the `ignored` option of chokidar. Again, from chokidar's documentation:
```
`ignored` [...] If a function with two arguments is provided, it gets called twice per path - once with a single argument (the path), second time with two arguments (the path and the fs.Stats object of that path).
```
then that function would use the dockeringore pattern matcher to tell chokidar whether or not to ignore the path.

It is not currently clear to me which approach would be more efficient -- consider both RAM and CPU. Keep in mind that a node_modules folder may contain tens of thousands of files (the CLI has some 50,000 files there!), so the an exhaustive list of files to watch could in some cases be smaller than the list of files to ignore...

Reply:
Oh, my first alternative approach would not work, because files newly created by the user (while livepush is running) would not be known to chokidar. I guess this leaves us with the second alternative approach - passing a function as the value of `ignored`.

Reply:
@pdcastro Thanks for the quick feedback, didn't think about the differences between the two pattern matchers. I'll try those ideas out and see how it goes. And JF now has multiple `node_modules` directories each with thousands of files so I agree setting watch instead of calling a func for ignored is probably going to be the cheaper solution.

Reply:
@pdcastro Updated the modifications to couple some existing logic to work with chokidar's ignored function. Let me know what you think.

Reply:
Have been using the build from this branch locally to keep working on getting Jellyfish to work with Livepush and have noticed that `balena push`ing to my local mode device sometimes exits without error once builds complete. I've added logs in an attempt to discern if this is somehow caused by my edits, but have not been able to figure out the root problem.